util/repr-bytes/src/lib.rs

```diff
@@ -0,0 +1,515 @@
+///   - <TryFrom<&[u8]> for $mytype>::Error implements core::fmt::Display
+///   - <TryFrom<&'a[u8]> for $mytype>::Error is the same for all values of 'a,
+///     OR they are all convertible to the value when 'a = 'static, via core::convert::From.
+#[macro_export]
+macro_rules! derive_repr_bytes_from_as_ref_and_try_from {
+    ($mytype:ty, $mysize:ty) => {
+        impl $crate::ReprBytes for $mytype {
+            type Size = $mysize;
+            type Error = <$mytype as ::core::convert::TryFrom<&'static [u8]>>::Error;
+
+            fn from_bytes(src: &$crate::GenericArray<u8, Self::Size>) -> Result<Self, Self::Error> {
+                Ok(<Self as ::core::convert::TryFrom<&[u8]>>::try_from(
+                    src.as_slice(),
+                )?)
+            }
+
+            fn to_bytes(&self) -> $crate::GenericArray<u8, Self::Size> {
+                use $crate::{typenum::Unsigned, GenericArray};
+                let slice = <Self as AsRef<[u8]>>::as_ref(self);
+                debug_assert!(slice.len() == <Self::Size as Unsigned>::USIZE);
+                <GenericArray<u8, Self::Size>>::from_slice(slice).clone()
+            }
+
+            // Optimization: Use AsRef<[u8]> here, which also makes Into<Vec<u8>> better
+            fn map_bytes<T, F>(&self, f: F) -> T
+            where
+                F: FnOnce(&[u8]) -> T,
+            {
+                f(<Self as AsRef<[u8]>>::as_ref(self))
+            }
+        }
+    };
+}
+
+/// Derive Into<Vec<u8>> from a ReprBytes implementation
+#[macro_export]
+macro_rules! derive_into_vec_from_repr_bytes {
+    ($mytype:ty) => {
+        impl Into<$crate::_exports::alloc::vec::Vec<u8>> for $mytype {
+            fn into(self) -> Vec<u8> {
+                <$mytype as $crate::ReprBytes>::map_bytes(&self, |slice| slice.to_vec())
+            }
+        }
+    };
+}
+
+/// Derive TryFrom<&[u8]> from a ReprBytes implementation
+/// Preconditions: ReprBytes::Error implements From<LengthMismatch>
+#[macro_export]
+macro_rules! derive_try_from_slice_from_repr_bytes {
+    ($mytype:ty) => {
+        impl<'a> ::core::convert::TryFrom<&'a [u8]> for $mytype {
+            type Error = <Self as $crate::ReprBytes>::Error;
+            fn try_from(src: &'a [u8]) -> Result<Self, Self::Error> {
+                if src.len() != <Self as $crate::ReprBytes>::size() {
+                    return Err(Self::Error::from($crate::LengthMismatch {
+                        expected: <Self as $crate::ReprBytes>::size(),
+                        found: src.len(),
+                    }));
+                }
+                <$mytype as $crate::ReprBytes>::from_bytes(<$crate::GenericArray<
+                    u8,
+                    <Self as $crate::ReprBytes>::Size,
+                >>::from_slice(src))
+            }
+        }
+    };
+}
+
+/// Derive prost::Message from a ReprBytes implementation
+/// The corresponding protobuf has exactly one member, of type `bytes`.
+#[macro_export]
```

Comment:
This should get gated on `cfg(feature = "prost")`, which I think would make this the only time we've been able to actually use a cargo feature as it was intended :-)